### PR TITLE
Functional Tests : GA for checking TypeScript

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Functional Tests : GA for checking TypeScript
| Type?             | improvement
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Relative to #30263 
| Related PRs       | N/A
| How to test?      | CI is :green_circle: : https://github.com/Progi1984/ga.tests.ui.pr/actions/runs/4893234855
| Possible impacts? | N/A

## Related PRs
* ashupp/js-image-generator#3
* ~~maildev/maildev#453 (Fixes maildev/maildev#445)~~

### CODE REVIEW

File: tests/UI/package.json [hunk 1]
```diff
@@ -9,6 +9,8 @@
     "// ESLint": "To run ESLint commands",
     "lint:fix": "eslint --fix --ignore-path .gitignore .",
     "lint": "eslint --ignore-path .gitignore .",
+    "// TypeScript Check": "Script to check typescript",
+    "check:typescript": "tsc --build ./tsconfig.json ",
```
Comment (0x346e3730) [inline tests/UI/package.json]: Add the `--noEmit` option to avoid generating the build if we only want to check for errors ?

**Review by 0x346e3730 — DISMISSED** [review]

File: tests/UI/package.json (hunk 1 above)
Comment (Progi1984) [reply tests/UI/package.json]: Sorry but it's not possible 

`Compiler option '--noEmit' may not be used with '--build'.`

File: tests/UI/tsconfig.json [hunk 2]
```diff
@@ -19,6 +19,8 @@
       "@utils/*": ["utils/*"],
       "@tools/*": ["tools/*"]
     },
-    "typeRoots": ["types", "node_modules/@types"]
+    "typeRoots": ["types", "node_modules/@types"],
+    "outDir": "dist",
+    "skipLibCheck": true
```
Comment (boubkerbribri) [inline tests/UI/tsconfig.json]:
```suggestion
    "typeRoots": ["types", "node_modules/@types"],
    "exclude": ["node_modules"],
    "noEmit": true
```

Comment (Progi1984) [reply tests/UI/tsconfig.json]: Or not : 
image

Comment (Progi1984) [reply tests/UI/tsconfig.json]: But the noEmit is :ok_hand:

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts [hunk 3]
```diff
@@ -131,19 +131,24 @@ describe('BO - Orders : Pagination of orders table', async () => {
       await testContext.addContextItem(this, 'testIdentifier', 'sortOrdersDesc', baseContext);
 
       const nonSortedTable = await ordersPage.getAllRowsColumnContent(page, 'total_paid_tax_incl');
+      const nonSortedTableFloat: number[] = nonSortedTable.map((text: string): number => parseFloat(text));
 
       await ordersPage.sortTable(page, 'total_paid_tax_incl', 'desc');
 
       sortedTable = await ordersPage.getAllRowsColumnContent(page, 'total_paid_tax_incl');
+      console.log(sortedTable);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts [hunk 4]
```diff
@@ -131,19 +131,24 @@ describe('BO - Orders : Pagination of orders table', async () => {
...
+      const expectedResult = await basicHelper.sortArrayNumber(nonSortedTableFloat);
+      await expect(sortedTableFloat).to.deep.equal(expectedResult.reverse());
     });
 
     it('should check that the orders table is sorted by total desc', async function () {
       await testContext.addContextItem(this, 'testIdentifier', 'CheckSortDesc', baseContext);
 
       const allOrdersTable = await ordersPage.getAllRowsColumnContent(page, 'total_paid_tax_incl');
+      console.log(sortedTable);
+      console.log(allOrdersTable);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts]:
```suggestion
```

File: tests/UI/campaigns/productV2/functional/01_filterProducts.ts [hunk 5]
```diff
@@ -191,7 +192,11 @@ describe('BO - Catalog - Products : Filter in Products Page', async () => {
 
           switch (test.args.comparisonType) {
             case 'toWithinMinMax':
-              await expect(textColumn).to.within(test.args.filterValue.min, test.args.filterValue.max);
+              console.log(test.args.filterValue);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/productV2/functional/01_filterProducts.ts]:
```suggestion
```

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts (hunk 3 above)
Comment (Progi1984) [reply tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts]: Fixed

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts (hunk 4 above)
Comment (Progi1984) [reply tests/UI/campaigns/functional/BO/02_orders/01_orders/10_pagination.ts]: Fixed

File: tests/UI/campaigns/productV2/functional/01_filterProducts.ts (hunk 5 above)
Comment (Progi1984) [reply tests/UI/campaigns/productV2/functional/01_filterProducts.ts]: Fixed

**Review by nesrineabdmouleh — DISMISSED** [review]

File: tests/UI/campaigns/sanity/06_checkoutFO/01_orderProduct.ts [hunk 6]
```diff
@@ -105,7 +105,7 @@ describe('BO - Checkout : Order a product and check order confirmation', async (
     const isStepPersonalInformationComplete = await checkoutPage.isStepCompleted(
       page,
       checkoutPage.personalInformationStepForm,
-    );
+    ); ;
```
Comment (boubkerbribri) [inline tests/UI/campaigns/sanity/06_checkoutFO/01_orderProduct.ts]:
```suggestion
    ); 
```

File: tests/UI/data/demo/countries.ts [hunk 7]
```diff
@@ -5,47 +5,47 @@ export default {
     id: 8,
     name: 'France',
     isoCode: 'FR',
-    callPrefix: 33,
+    callPrefix: '33',
```
Comment (boubkerbribri) [inline tests/UI/data/demo/countries.ts:8]: Why it's not a number ?

File: tests/UI/pages/BO/catalog/products/add.ts [hunk 8]
```diff
@@ -273,7 +273,7 @@ class AddProduct extends BOBasePage {
    * @returns {Promise<void>}
    */
   async addProductImages(page: Page, imagesPaths: (string | null)[] = []): Promise<void> {
-    const filteredImagePaths: string[] = imagesPaths.filter((el: string | null) => el !== null);
+    const filteredImagePaths: string[] = imagesPaths.filter((el: string | null): el is string => el !== null);
```
Comment (boubkerbribri) [inline tests/UI/pages/BO/catalog/products/add.ts:276]: what do you mean by el is string ?

File: tests/UI/pages/BO/orders/add.ts [hunk 9]
```diff
@@ -9,6 +9,7 @@ import type OrderStatusData from '@data/faker/orderStatus';
 
 import type {Frame, Page} from 'playwright';
 import OrderData from '@data/faker/order';
+import {expect} from 'chai';
```
Comment (boubkerbribri) [inline tests/UI/pages/BO/orders/add.ts]: expect shouldn't be used on a page.

File: tests/UI/tsconfig.json [hunk 10]
```diff
@@ -19,6 +19,9 @@
       "@utils/*": ["utils/*"],
       "@tools/*": ["tools/*"]
     },
-    "typeRoots": ["types", "node_modules/@types"]
+    "typeRoots": ["types", "node_modules/@types"],
+    "outDir": "dist",
```
Comment (boubkerbribri) [inline tests/UI/tsconfig.json]: do you need a `outDir` when we have noEmit ?

File: tests/UI/campaigns/sanity/06_checkoutFO/01_orderProduct.ts (hunk 6 above)
Comment (Progi1984) [reply tests/UI/campaigns/sanity/06_checkoutFO/01_orderProduct.ts]: Fixed

File: tests/UI/data/demo/countries.ts (hunk 7 above)
Comment (Progi1984) [reply tests/UI/data/demo/countries.ts:8]: Because, it can contains a `+` in some cases.

File: tests/UI/pages/BO/catalog/products/add.ts (hunk 8 above)
Comment (Progi1984) [reply tests/UI/pages/BO/catalog/products/add.ts:276]: It's a user-defined type guard.  

https://stackoverflow.com/a/45748366

File: tests/UI/pages/BO/orders/add.ts (hunk 9 above)
Comment (Progi1984) [reply tests/UI/pages/BO/orders/add.ts]: Fixed

File: tests/UI/tsconfig.json (hunk 10 above)
Comment (Progi1984) [reply tests/UI/tsconfig.json]: Fixed

**Review by nesrineabdmouleh — APPROVED** [review]

**Review by boubkerbribri — APPROVED** [review]

**Review by mflasquin — APPROVED** [review]

**Review by mflasquin — APPROVED** [review]

**Comment by boubkerbribri** [comment]
Thanks @Progi1984

**Pull request merged**